pkg/util/pki/csr_test.go

```diff
@@ -562,6 +590,53 @@ func TestGenerateCSR(t *testing.T) {
 				ExtraExtensions:    ipsecExtraExtensions,
 			},
 		},
+		{
+			name: "Generate CSR from certificate with a single otherNameSAN set to an oid (UPN)", // only a shallow validation is expected
+			crt: &cmapi.Certificate{Spec: cmapi.CertificateSpec{OtherNameSANs: []cmapi.OtherNameSAN{
+				{
+					Type:        "oid:1.3.6.1.4.1.311.20.2.3",
+					StringValue: "user@example.org",
+				},
+			}}},
+			want: &x509.CertificateRequest{
+				Version:            0,
+				SignatureAlgorithm: x509.SHA256WithRSA,
+				PublicKeyAlgorithm: x509.RSA,
+				ExtraExtensions:    append(defaultExtraExtensions, asn1otherNameUpnSANExpectedExtension),
+			},
+		},
+		{
+			name: "Generate CSR from certificate with multiple valid otherName oids and emailSANs set",
+			crt: &cmapi.Certificate{Spec: cmapi.CertificateSpec{
+				EmailAddresses: []string{"user@example.org", "alt-email@example.org"},
+				OtherNameSANs: []cmapi.OtherNameSAN{
+					{
+						Type:        "oid:1.3.6.1.4.1.311.20.2.3",
+						StringValue: "user@example.org",
+					},
+					{
+						Type:        "oid:1.2.840.113556.1.4.221", // this is the legacy sAMAccountName to make example 'realistic' but could be any oid
+						StringValue: "user@example.org",
+					},
+				}}},
+			want: &x509.CertificateRequest{
+				Version:            0,
+				SignatureAlgorithm: x509.SHA256WithRSA,
+				PublicKeyAlgorithm: x509.RSA,
+				ExtraExtensions:    append(defaultExtraExtensions, asn1otherNameUpnSANExpectedExtension, asn1otherNamesAMAAccountNameExpectedExtension),
+				EmailAddresses:     []string{"user@example.org", "alt-email@example.org"},
```

Comment:
When you encode this `x509.CertificateRequest` object, the encoded version will only contain the UPN values, the EmailAddresses will be skipped: https://cs.opensource.google/go/go/+/refs/tags/go1.21.2:src/crypto/x509/x509.go;l=1166-1167